iotfunctions/anomaly.py

```diff
@@ -1339,6 +1320,93 @@ def get_input_items(cls):
         return ['features', 'targets']
 
 
+class SimpleRegressor(BaseEstimatorFunction):
+
+    '''
+    Regressor based on stochastic gradient descent and gradient boosting method as provided by sklearn
+    '''
+    eval_metric = staticmethod(metrics.r2_score)
+
+    # class variables
+    train_if_no_model = True
+    estimators_per_execution = 3
+    num_rounds_per_estimator = 3
+
+    def GBRPipeline(self):
+        steps = [('scaler', StandardScaler()), ('gbr', ensemble.GradientBoostingRegressor)]
+        return Pipeline(steps)
+
+    def SGDPipeline(self):
+        steps = [('scaler', StandardScaler()), ('sgd', linear_model.SGDRegressor)]
+        return Pipeline(steps)
+
+    def set_estimators(self):
+        # gradient_boosted
+        params = {'gbr__n_estimators': [100, 250, 500, 1000],
+                  'gbr__max_depth': [2, 4, 10],
+                  'gbr__min_samples_split': [2, 5, 9],
+                  'gbr__learning_rate': [0.01, 0.02, 0.05],
+                  'gbr__loss': ['ls']}
+        # self.estimators['gradient_boosted_regressor'] = (ensemble.GradientBoostingRegressor, params)
+        self.estimators['gradient_boosted_regressor'] = (self.GBRPipeline, params)
+
+        # sgd
+        params = {'sgd__max_iter': [250, 1000, 5000, 10000],
+                  'sgd__tol': [0.001, 0.002, 0.005]}
+        # self.estimators['sgd_regressor'] = (linear_model.SGDRegressor, params)
+        self.estimators['sgd_regressor'] = (self.SGDPipeline, params)
+        logger.info('SimpleRegressor start searching for best model')
+
+    def __init__(self, features, targets, predictions=None,
+                 n_estimators=None, num_leaves=None, learning_rate=None, max_depth=None):
+        super().__init__(features=features, targets=targets, predictions=predictions)
+
+        self.experiments_per_execution = 1
+        self.auto_train = True
+
+        # self.stop_auto_improve_at = -2
+
+    def execute(self, df):
+
+        df_copy = df.copy()
+        entities = np.unique(df_copy.index.levels[0])
```

Comment:
is it entity ids ? what happens if the level changes from 0 to something else ?  Better to use index names.

Reply:
I'm not aware that there is a documented naming convention for entities and timestamps.
All the anomaly scorers rely on level 0 being the level of entity and I extended this approach to SimpleRegressor.